dartagnan/src/main/java/com/dat3m/dartagnan/expression/INonDet.java

```diff
@@ -136,21 +145,19 @@ public int getPrecision() {
 		return precision;
 	}
 	
-	public BoolExpr encodeBounds(boolean bp, Context ctx) {
-		BoolExpr enc = ctx.mkTrue();
+	public BooleanFormula encodeBounds(boolean bp, SolverContext ctx) {
+		FormulaManager fmgr = ctx.getFormulaManager();
+		BooleanFormulaManager bmgr = fmgr.getBooleanFormulaManager();
+		BooleanFormula enc = bmgr.makeTrue();
 		long min = getMin();
 		long max = getMax();
 		if(bp) {
-			if(type.equals(UINT) || type.equals(ULONG) || type.equals(USHORT) || type.equals(UCHAR)) {
-	        	enc = ctx.mkAnd(enc, ctx.mkBVUGE((BitVecExpr)toZ3Int(null,ctx), ctx.mkBV(min, precision)));
-	        	enc = ctx.mkAnd(enc, ctx.mkBVULE((BitVecExpr)toZ3Int(null,ctx), ctx.mkBV(max, precision)));					
-			} else {
-	        	enc = ctx.mkAnd(enc, ctx.mkBVSGE((BitVecExpr)toZ3Int(null,ctx), ctx.mkBV(min, precision)));
-	        	enc = ctx.mkAnd(enc, ctx.mkBVSLE((BitVecExpr)toZ3Int(null,ctx), ctx.mkBV(max, precision)));					
-			}
+			boolean signed = !(type.equals(UINT) || type.equals(ULONG) || type.equals(USHORT) || type.equals(UCHAR));
+			enc = bmgr.and(enc, fmgr.getBitvectorFormulaManager().greaterOrEquals((BitvectorFormula) toIntFormula(null,ctx), fmgr.getBitvectorFormulaManager().makeBitvector(precision, min), signed));
+	        enc = bmgr.and(enc, fmgr.getBitvectorFormulaManager().lessOrEquals((BitvectorFormula) toIntFormula(null,ctx), fmgr.getBitvectorFormulaManager().makeBitvector(precision, max), signed));
 		} else {
-        	enc = ctx.mkAnd(enc, ctx.mkGe((IntExpr)toZ3Int(null,ctx), ctx.mkInt(min)));
-        	enc = ctx.mkAnd(enc, ctx.mkLe((IntExpr)toZ3Int(null,ctx), ctx.mkInt(max)));
+			enc = bmgr.and(enc, fmgr.getIntegerFormulaManager().greaterOrEquals((IntegerFormula)toIntFormula(null,ctx), fmgr.getIntegerFormulaManager().makeNumber(min)));
+			enc = bmgr.and(enc, fmgr.getIntegerFormulaManager().lessOrEquals((IntegerFormula)toIntFormula(null,ctx), fmgr.getIntegerFormulaManager().makeNumber(max)));
```

Comment:
Refactor fmgr.getIntegerFormulaManager / fmgr.getBitVectorFormulaManager into local variables of their respective branches.

Reply:
About to be pushed